org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Scanner.java

```diff
@@ -2032,99 +1862,129 @@ protected int processSingleQuotes(boolean checkIfUnicode) throws InvalidInputExc
+							int start = this.currentPosition;
+							for (int lookAhead = 0; lookAhead < 50; lookAhead++) {
+								if (this.currentPosition >= this.eofPosition) {
+									this.currentPosition = start;
+									break;
+								}
+								if (((this.currentCharacter = this.source[this.currentPosition++]) == '\\') && (this.source[this.currentPosition] == 'u')) {
+									isUnicode = true;
+									getNextUnicodeChar();
+								} else {
+									isUnicode = false;
+								}
+								if (!isUnicode && this.currentCharacter == '\n') {
+									this.currentPosition--; // set current position on new line character
+									break;
+								}
+								if (this.currentCharacter == '\"') {
+									throw new InvalidInputException(INVALID_CHAR_IN_STRING);
+								}
 							}
+						} else {
+							this.currentPosition--; // set current position on new line character
+						}
+						throw new InvalidInputException(INVALID_CHAR_IN_STRING);
+					case '}':
+						if (openBraces > 0 && --openBraces == 0) {
+							pushBracePosition(this.currentPosition - (isUnicode ? 6 : 1));
+							pushFragmentPosition(this.withoutUnicodePtr, this.currentPosition - 1);
+						}
+						break;
+					case '{':
+						if (openBraces > 0) {
+							openBraces++;
+						}
+						break;
+					case '\\':
+						if (this.unicodeAsBackSlash) {
+							this.withoutUnicodePtr--;
+							// consume next character
+							this.unicodeAsBackSlash = false;
 							if (((this.currentCharacter = this.source[this.currentPosition++]) == '\\') && (this.source[this.currentPosition] == 'u')) {
-								isUnicode = true;
 								getNextUnicodeChar();
+								isUnicode = true;
+								this.withoutUnicodePtr--;
 							} else {
 								isUnicode = false;
 							}
-							if (!isUnicode && this.currentCharacter == '\n') {
-								this.currentPosition--; // set current position on new line character
-								break;
-							}
-							if (this.currentCharacter == '\"') {
-								throw invalidCharInString();
+						} else {
+							if (this.withoutUnicodePtr == 0) {
+								unicodeInitializeBuffer(this.currentPosition - this.startPosition);
 							}
+							this.withoutUnicodePtr --;
+							this.currentCharacter = this.source[this.currentPosition++];
 						}
-					} else {
-						this.currentPosition--; // set current position on new line character
-					}
-					throw invalidCharInString();
-				}
-				if (this.currentCharacter == '\\') {
-					if (this.unicodeAsBackSlash) {
-						this.withoutUnicodePtr--;
-						// consume next character
-						this.unicodeAsBackSlash = false;
-						if (((this.currentCharacter = this.source[this.currentPosition++]) == '\\') && (this.source[this.currentPosition] == 'u')) {
-							getNextUnicodeChar();
-							isUnicode = true;
-							this.withoutUnicodePtr--;
+
+						if (templatesAllowed && this.currentCharacter == '{' && openBraces++ == 0) {
+							pushFragmentPosition(this.withoutUnicodePtr, this.currentPosition - (isUnicode ? 7 : 2));
```

Comment:
I think using jumpOverMethodBody can significantly simplify the implementation. Let us discuss